src/NHibernate/Impl/DelayedAsyncEnumerator.cs

```diff
@@ -30,39 +30,40 @@ IAsyncEnumerator<T> IAsyncEnumerable<T>.GetEnumerator()
 
 		private sealed class Enumerator : IAsyncEnumerator<T>
 		{
-			private IEnumerable data;
-			private GetResult result;
-			private Delegate executeOnEval;
-			private IEnumerator enumerator;
```

Comment:
Same in synchronous version. The trouble is, old `IEnumerator` is not `IDisposable`, but we indeed affect it a `IEnumerator<T>` which is and was not disposed. I have fix this new async version, and intend to add a Jira for fixing current synchronous code.